File: renewal_module/renewal_module/report/test__button/test__button.py
# Copyright (c) 2023, Aravind Mandala and contributors
# For license information, please see license.txt

# import frappe


# def execute(filters=None):
# 	columns, data = [], []
# 	return columns, data


import frappe

def execute(filters=None):
    return get_columns(), get_data()

def get_data():
    current_user = frappe.session.user
    todos = frappe.get_all(
        "ToDo",
        filters={"status": "Open"},
        # created by or allocated to current user
        or_filters=[{"owner": current_user}, {"allocated_to": current_user}],
        fields=["name", "description", "status"],
    )
    return todos

# 3 columns
def get_columns():
    return [
        {
            "label": "Description",
            "fieldname": "description",
            "fieldtype": "Data",
            "width": 400,
        },
        {"label": "Status", "fieldname": "status", "fieldtype": "Data", "width": 200},

        # This column will be used to render the close buttons
        {
            "label": "Close Action",
            "fieldname": "name",
            "fieldtype": "Data",
            "width": 150,
        },
    ]

File: renewal_module/renewal_module/report/test__button/test__button.js
// Copyright (c) 2023, Aravind Mandala and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Test- Button"] = {
	"filters": [],
  onload(report) {
    // add an action button to visit ToDo List View
    report.page.add_inner_button("Go to ToDo List", () => {
      frappe.set_route("List", "ToDo");
    });
  },
  formatter(value, row, column, data, default_formatter) {
    // Show a button instead of the "name"
    if (column.fieldname == "name") {
        const button_html = `<button class="btn btn-default btn-xs" onclick="frappe.query_reports['Test- Button'].close_todo('${value}')">Close</button>`;
        value = button_html;
    }
    return default_formatter(value, row, column, data);
  },
  close_todo(name) {
    frappe.db.set_value("ToDo", name, "status", "Closed").then(() => {
      // refresh this report and show alert
      frappe.query_report.refresh();
      frappe.show_alert("ToDo Closed Successfully!");
    });
  },
};
